### PR TITLE
Refactor pipeline and listeners to propagate all Transformer outputs

### PR DESCRIPTION
**Pull request by shadeMe**
Models that follow the transformer in the pipeline model chain operate on the outputs of all the transformer layers. This will allow for the usage of a separate listener that can pass a weighted sum of the former to the pooling layer (as opposed to using just the last layer's hidden state).

### CODE REVIEW

File: curated_transformers/models/__init__.py [hunk 1]
```diff
@@ -1,5 +1,11 @@
+from typing import Union
```
Comment (danieldk) [inline curated_transformers/models/__init__.py:1]: Do we need this import here?

File: curated_transformers/models/hf_wrapper.py [hunk 2]
```diff
@@ -72,24 +70,30 @@ def convert_from_torch_backward(d_inputs):
...
+    Yt = [
+        [output[i, :len, :] for output in model_outputs.all_outputs]
+        for i, len in enumerate(input_lens)
+    ]
+    Y = [[torch2xp(layer, ops=ops) for layer in output] for output in Yt]
+    output = TransformerModelOutput(outputs=Y)
+
+    def convert_for_torch_backward(dY: List[List[Floats2d]]):
+        Yt_flat = [y for inner in Yt for y in inner]
+        dYt_flat = [xp2torch(y) for inner in dY for y in inner]
```
Comment (danieldk) [inline curated_transformers/models/hf_wrapper.py:88]: I think this already came up in an earlier discussion -- we'd always backprop through all layers, even if we only use last layer listeners. The performance impact may not be that bad, since the gradients are accumulated, but did you have a chance to measure the difference?

File: curated_transformers/models/output.py [hunk 3]
```diff
@@ -26,9 +30,53 @@ def layer_hidden_state(self, idx: int) -> Tensor:
...
+    Also acts as the accumulator for the outputs of subsequent models in the Transformer pipeline."""
+
+    # Non-padded, un-stacked versions of the outputs.
+    # The outer list tracks Docs/spans and the inner
+    # list tracks the embedding + layer outputs of each Doc/span.
+    #
+    # The inner-most element is a Floats2d when returned by
+    # the PyTorchWrapper transformer model, which are subsequently
+    # converted to Ragged by the models that follow.
+    all_outputs: List[List[Union[Floats2d, Ragged]]]
```
Comment (danieldk) [inline curated_transformers/models/output.py:62]: I wonder if we can make this a typevar with bounds rather than a union. Then we can reason about what is inside this container and e.g. `embedding_outputs` would return the correct type. Or is it actually a mixture. If not, and a typevar is not an option, the union should be the outermost type.

File: curated_transformers/models/with_strided_spans.py [hunk 4]
```diff
@@ -1,16 +1,24 @@
-from typing import Callable, List, Tuple
+from typing import Callable, List, Tuple, Union
 from functools import partial
+from unittest.mock import sentinel
```
Comment (danieldk) [inline curated_transformers/models/with_strided_spans.py:3]: Stray import?

File: curated_transformers/models/with_strided_spans.py [hunk 5]
```diff
@@ -56,6 +66,14 @@ def with_strided_spans_forward(
 
     Y_layer, backprop_layer = model.layers[0](spans, is_train=is_train)
 
+    is_trf_output = isinstance(Y_layer, TransformerModelOutput)
```
Comment (danieldk) [inline curated_transformers/models/with_strided_spans.py:69]: Could you explain this conditional block? In what case it the layer output not transformer output?

File: curated_transformers/models/with_strided_spans.py [hunk 6]
```diff
@@ -1,16 +1,24 @@
...
+
 
 def build_with_strided_spans_v1(stride=96, window=128):
     return partial(with_strided_spans, stride=stride, window=window)
 
 
 def with_strided_spans(
     layer, *, stride=96, window=128
-) -> Model[List[Ragged], List[Ragged]]:
+) -> Model[List[Ragged], Union[TransformerModelOutput, RaggedInOutT]]:
```
Comment (danieldk) [inline curated_transformers/models/with_strided_spans.py:21]: This ties in with the other comment. But I don't really understand why we need the unions here. Aren't we always dealing with transformer outputs?

File: curated_transformers/pipe.py [hunk 7]
```diff
@@ -258,13 +262,15 @@ def predict(self, docs: Iterable[Doc]):
...
+    def set_annotations(
+        self, docs: Sequence[Doc], trf_output: TransformerModelOutput
+    ) -> None:
         """Modify a batch of documents, using pre-computed scores.
         docs (Iterable[Doc]): The documents to modify.
         tokvecses: The tensors to set, produced by Tok2Vec.predict.
         DOCS: https://spacy.io/api/tok2vec#set_annotations
         """
-        for doc, tokvecs in zip(docs, tokvecses):
+        for doc, tokvecs in zip(docs, trf_output.all_outputs):
```
Comment (danieldk) [inline curated_transformers/pipe.py:273]: I wonder if we can do nicer here, by wrapping the per-doc outputs in another class. E.g. `DocTransformerModelOutput` (there is probably a better name). Then we don't need the `-1` indexing in other places, but could just have a property to access the last hidden layer.

File: curated_transformers/tokenization/sentencepiece_adapters.py [hunk 8]
```diff
@@ -10,6 +12,8 @@
 _SPP_EOS = 2
 _SPP_UNK = 0
 
+RemoveBosEosInOutT = Union[List[Ragged], List[List[Ragged]], TransformerModelOutput]
```
Comment (danieldk) [inline curated_transformers/tokenization/sentencepiece_adapters.py:15]: Same question about this type, don't we always have `TransformerModelOutput`?

File: curated_transformers/models/__init__.py (hunk 1 above)
Comment (shadeMe) [reply curated_transformers/models/__init__.py:1]: Seems like a leftover, will remove.

File: curated_transformers/models/hf_wrapper.py (hunk 2 above)
Comment (shadeMe) [reply curated_transformers/models/hf_wrapper.py:88]: Not yet, but I can train a model to check before we merge this PR.

File: curated_transformers/models/output.py (hunk 3 above)
Comment (shadeMe) [reply curated_transformers/models/output.py:62]: It's not a mixture in that all the innermost elements are of the same time at any given point, i.e., either a `Ragged` or a `Floats2d`. In fact, they are the former in all cases except inside the `with_strided_spans` model.

File: curated_transformers/models/with_strided_spans.py (hunk 4 above)
Comment (shadeMe) [reply curated_transformers/models/with_strided_spans.py:3]: Yup, will remove.

File: curated_transformers/models/with_strided_spans.py (hunk 5 above)
Comment (shadeMe) [reply curated_transformers/models/with_strided_spans.py:69]: At the moment, some of the tests for the `with_strided_spans` model simply pass in `Ragged` inputs. I thought it would be cleaner to support them instead of wrapping them in a `TransformerModelOutput` (and it makes the model usable outside of transformer contexts).

File: curated_transformers/models/with_strided_spans.py (hunk 6 above)
Comment (shadeMe) [reply curated_transformers/models/with_strided_spans.py:21]: As mentioned above, we support inner layers that either return a `TransformerModelOutput` or a (nested) list of `Ragged`s in the current implementation.

File: curated_transformers/pipe.py (hunk 7 above)
Comment (shadeMe) [reply curated_transformers/pipe.py:273]: Sounds good, I'll add that.

File: curated_transformers/tokenization/sentencepiece_adapters.py (hunk 8 above)
Comment (shadeMe) [reply curated_transformers/tokenization/sentencepiece_adapters.py:15]: Same answer as above.

File: curated_transformers/models/output.py (hunk 3 above)
Comment (danieldk) [reply curated_transformers/models/output.py:62]: So this could be `Union[List[List[Floats2d]], List[List[Ragged]]]` or even a type variable with bounds?

File: curated_transformers/models/with_strided_spans.py (hunk 5 above)
Comment (danieldk) [reply curated_transformers/models/with_strided_spans.py:69]: I think we should initially just make `with_strided_spans` work with `TransformerModelOutput`s. I think it's better to make the tests mock some `MakeModelOutputs` than to have union types and condition blocks to handle them this early on. It makes the types harder to read, the implementations more difficult to follow, and I think it is unlikely that these layers would be used outside `curated-transformers`.

File: curated_transformers/models/output.py (hunk 3 above)
Comment (shadeMe) [reply curated_transformers/models/output.py]: I've added a type variable for this.

File: curated_transformers/models/hf_wrapper.py (hunk 2 above)
Comment (shadeMe) [reply curated_transformers/models/hf_wrapper.py:101]: The performance impact of processing all layer outputs is significant (~20-23% reduction in WPS), but the latest changes should restrict the penalty to only those cases where all outputs are used downstream. When they aren't, we simply remove all but the last hidden representation from the `TransformerOutput` instance, thereby avoiding redundant processing.

File: curated_transformers/models/hf_wrapper.py [hunk 9]
```diff
@@ -36,6 +35,10 @@ def build_hf_transformer_encoder_v1(
         grad_scaler=PyTorchGradScaler(**grad_scaler_config),
     )
 
+    # Managed by the parent Pipe instance. Remove all but the last
+    # layer outputs if the rest aren't required by the downstream models.
+    model.attrs["all_layer_outputs"] = True
```
[the remark this answers is not in the document]
Comment (shadeMe) [reply curated_transformers/models/hf_wrapper.py:40]: After internal discussion, we've decided to implement this as a pipe-level, set-and-forget option.

File: curated_transformers/pipe.py [hunk 10]
```diff
@@ -207,6 +225,14 @@ def add_listener(
         self, listener: "TransformerListener", component_name: str
     ) -> None:
         """Add a listener for a downstream component. Usually internals."""
+        if (
+            not isinstance(listener, LastTransformerLayerListener)
```
Comment (danieldk) [inline curated_transformers/pipe.py:229]: I think this check should be in the listener, since there can be other custom listeners that only use the last layer.

Comment (shadeMe) [reply curated_transformers/pipe.py]: Added `needs_all_layer_outputs` to `TransformerListener`.

**Review by danieldk — APPROVED** [review]
